akka-javasdk/src/main/scala/akka/javasdk/impl/consumer/ReflectiveConsumerRouter.scala

```diff
@@ -23,14 +25,16 @@ private[impl] class ReflectiveConsumerRouter[A <: Consumer](
     consumer: A,
     methodInvokers: Map[String, MethodInvoker],
     serializer: JsonSerializer,
-    ignoreUnknown: Boolean)
-    extends ConsumerRouter[A](consumer) {
+    ignoreUnknown: Boolean) {
 
-  override def handleUnary(message: MessageEnvelope[BytesPayload]): Consumer.Effect = {
+  def handleCommand(message: MessageEnvelope[BytesPayload], context: MessageContext): Consumer.Effect = {
+    // only set, never cleared, to allow access from other threads in async callbacks in the consumer
+    // the same handler and consumer instance is expected to only ever be invoked for a single message
+    consumer._internalSetMessageContext(Optional.of(context))
 
     val payload = message.payload()
     // make sure we route based on the new type url if we get an old json type url message
-    val inputTypeUrl = serializer.removeVersion(AnySupport.replaceLegacyJsonPrefix(payload.contentType))
+    val inputTypeUrl = serializer.removeVersion(serializer.replaceLegacyJsonPrefix(payload.contentType))
```

Comment:
consumer is different from the others, but I guess there is a reason, such as external messages from broker